### DISCUSSION TITLE
Usage with Expo

### DISCUSSION THREAD

**Discussion by Tobbe** [kirillzyusko/react-native-keyboard-controller#729] [Q&A] [answered]
Hi!

This package looks really promising, but since I'm new to React Native development with Expo I need some help to get it working.

I run `eas build --profile development --platform android --local` to get a development build of my app. I then drag-and-drop the resulting .apk to my Android emulator. The app starts and asks to connect an Expo dev server. So I start one up with `npx expo start`, and then I can connect and start the app.

But as soon as I make a change that requires me to reload the app I get this error

```
 ERROR  Error: The package 'react-native-keyboard-controller' doesn't seem to be linked. Make sure:

- You have run 'pod install'
- You rebuilt the app after installing the package
- You are not using Expo Go
, js engine: hermes [Component Stack]
```

Is there anyway I can avoid having to do a full rebuild of the app all the time?

**Comment by kirillzyusko** [comment]
Hello @Tobbe 

This is really strange that you get this error. I'd like to ask you few questions:
- do you use expo dev client? Not expo go?
- have you ever tried to build for iOS? Does it work there?
- have you tried to run `npm run android`? It should automatically assemble app and install it on your connected device

Overall you can find many expo apps that are using `keyboard-controller` and compare a setup. For example https://github.com/kimak/keyboardAwareScrollFormSheet/ could be used as a reference (i recently investigated why one component doesn't work, and had this reproduction project, so thought it could be helpful for you to have a look as well).

Also, I believe that you need to run `prebuild` [command](https://docs.expo.dev/workflow/prebuild/) if you add a new dependency with a native code (but `npm run android` should do this for you).

Let me know if prebuild/npm run android helped you or not 😊

**Reply by Tobbe** [reply]
@kirillzyusko Thanks for getting back to me 🙏

> do you use expo dev client? Not expo go?

Yes, I'm using the dev client. I'm still on the old architecture, and Expo Go now only works with the new architecture. So couldn't use it even if I wanted to 🙂

> have you tried to run npm run android? It should automatically assemble app and install it on your connected device
> [...]
> Also, I believe that you need to run prebuild [command](https://docs.expo.dev/workflow/prebuild/) if you add a new dependency with a native code (but npm run android should do this for you).

I would like to avoid prebuilding if possible. And I thought the dev build would be the way to avoid that workflow.

> have you ever tried to build for iOS? Does it work there?

Someone tried to help me on the Expo Discord, and you might be onto something here. I'm investigating right now. But if I had paid closer attention to the error message I would have noticed it said `You have to run 'pod install'`. Pods are for iOS only, not Android. So it seems the Expo Dev server is trying to do something for iOS even though I'm only running in the Android Simulator.

I'm trying an iOS build right now, and will reply back as soon as I have more info.

**Reply by Tobbe** [reply]
Also doing a local build for iOS got rid of the error! Yay! Now I can continue exploring this package 🎉

(I also uninstalled `react-native-keyboard-controller` as I had installed it with npm, and reinstalled it with `npx expo install react-native-keyboard-controller`. I'm not sure if that's relevant or not)

**Reply by Tobbe** [reply, marked as answer]
And for completeness, this is what the relevant part of my `eas.json` file looks like

```json
{
  "cli": {
    "version": ">= 12.5.1",
    "appVersionSource": "remote"
  },
  "build": {
    "development": {
      "developmentClient": true,
      "distribution": "internal",
      "channel": "development",
      "ios": {
        "simulator": true
      }
    },
```

**Reply by kirillzyusko** [reply]
> But if I had paid closer attention to the error message I would have noticed it said You have to run 'pod install'. Pods are for iOS only, not Android. So it seems the Expo Dev server is trying to do something for iOS even though I'm only running in the Android Simulator.

Oh, very good catch! On Android this error message should be present without "pod installation" instructions. Here I add it conditionally:

https://github.com/kirillzyusko/react-native-keyboard-controller/blob/80d89722083922c411b6501263b6e9835b4ee813/src/bindings.native.ts#L15

So after assembling iOS the problem has been solved for Android as well? Or Android is still broken? 🤔

**Reply by Tobbe** [reply]
Both work! 👍

**Reply by kirillzyusko** [reply]
Awesome! So where the problem was? 🤔 Did you just assemble iOS app and the error disappeared on Android too?

**Reply by Tobbe** [reply]
Yes, that. And I also used `npx expo install` instead of `npm install` to install react-native-keyboard-controller. Not sure if that was actually needed or not though.

**Reply by kirillzyusko** [reply]
> npx expo install instead of npm install

I think it did a trick 🙂

**Comment by kirillzyusko** [comment]
Close the discussion, because we found a solution 🚀
